src/Lib.hs

```diff
@@ -0,0 +1,228 @@
+import           Data.Text                                ( Text )
+import qualified Data.Text                     as T
+import           Data.Text.Encoding                       ( encodeUtf8 )
+import           Network.Connection                       ( Connection
+                                                          , ConnectionParams(..)
+                                                          , connectionGetChunk'
+                                                          , connectionPut
+                                                          , connectTo
+                                                          , initConnectionContext
+                                                          )
+import           Network.HTTP.Req                         ( GET(..)
+                                                          , NoReqBody(..)
+                                                          , Option
+                                                          , Url
+                                                          , defaultHttpConfig
+                                                          , lbsResponse
+                                                          , parseUrl
+                                                          , req
+                                                          , responseBody
+                                                          , runReq
+                                                          )
+import           Network.Socket                           ( PortNumber )
+import           System.IO                                ( hFlush
+                                                          , stdout
+                                                          )
+import           Text.HTML.TagSoup                        ( isTagCloseName
+                                                          , isTagOpenName
+                                                          , maybeTagText
+                                                          , parseTags
+                                                          )
+
+data Config = Config
+  { server   :: Text
+  , port     :: PortNumber
+  , chan     :: Text
+  , nick     :: Text
+  , password :: Maybe Text
+  , longname :: Text
+  }
+
+emptyConfig = Config
+  { server   = ""
+  , port     = 1024
+  , chan     = ""
+  , nick     = ""
+  , password = Nothing
+  , longname = ""
+  }
+
+data Bot = Bot
+  { connection   :: Connection
+  , config       :: Config
+  , isHandshaked :: MVar ()
+  }
+
+data Command
+  = ReplyWelcome B8.ByteString
+  | Ping B8.ByteString
+  | Privmsg B8.ByteString B8.ByteString B8.ByteString -- ^ Sender, Target, message
+  | Other B8.ByteString
+  deriving (Eq, Show)
+
+-- | Write a ByteString to the connection.
+writeRaw :: Connection -> B8.ByteString -> IO ()
+writeRaw connection t =
+  let t' = t <> "\r\n" in connectionPut connection t' *> B8.putStr t'
```

Comment:
I'd probably written it like this (of course it's alright to keep the let-clause):

```suggestion
  connectionPut connection t *> B8.putStr $ t <> endOfLine
    where endOfLine = "\r\n"
```

This way it's more clear why we're appending "\r\n" to the bytestring.

Reply:
Yeah, but I would probably rather add a comment to the function that all IRC-commands are ended with \r\n. I'm not much for indirection.